# Módulo 07, Tarefa 01

Vamos começar a mexer na nossa base de projeto? Já fizemos um exercício de montar a variável resposta, já pudemos perceber que essa atividade pode não ser trivial. Vamos agora trabalhar a base para que fique propícia ao *scikitlearn* para trabalharmos.

Lembrando, a base se chama demo01.csv, e originalmente está publicada [aqui](https://www.kaggle.com/rikdifos/credit-card-approval-prediction).

#### 1) Carregue a base e avalie:

- As variáveis
- Tipos de dados de cada variável
- Quantidade de missings
- Distribuição da variável resposta (mau)

In [45]:
import pandas as pd
import numpy as np


In [46]:
#carregando a base de dados
df = pd.read_csv('demo01.csv', sep=',')
df.head()

sexo posse_de_veiculo posse_de_imovel  qtd_filhos            tipo_renda  \
0    M                Y               Y           0               Working   
1    F                N               Y           0  Commercial associate   
2    F                N               Y           0  Commercial associate   
3    M                Y               Y           0               Working   
4    F                Y               N           0               Working   

                        educacao          estado_civil    tipo_residencia  \
0  Secondary / secondary special               Married  House / apartment   
1  Secondary / secondary special  Single / not married  House / apartment   
2  Secondary / secondary special  Single / not married  House / apartment   
3               Higher education               Married  House / apartment   
4              Incomplete higher               Married  House / apartment   

       idade  tempo_emprego  possui_celular  possui_fone_comercial  \
0  58.832877       3.106849               1                      0   
1  52.356164       8.358904               1                      0   
2  52.356164       8.358904               1                      0   
3  46.224658       2.106849               1                      1   
4  29.230137       3.021918               1                      0   

   possui_fone  possui_email  qt_pessoas_residencia    mau  
0            0             0                    2.0  False  
1            1             1                    1.0  False  
2            1             1                    1.0  False  
3            1             1                    2.0  False  
4            0             0                    2.0  False

In [47]:
#As variáveis
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16650 entries, 0 to 16649
Data columns (total 16 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   sexo                   16650 non-null  object 
 1   posse_de_veiculo       16650 non-null  object 
 2   posse_de_imovel        16650 non-null  object 
 3   qtd_filhos             16650 non-null  int64  
 4   tipo_renda             16650 non-null  object 
 5   educacao               16650 non-null  object 
 6   estado_civil           16650 non-null  object 
 7   tipo_residencia        16650 non-null  object 
 8   idade                  16650 non-null  float64
 9   tempo_emprego          16650 non-null  float64
 10  possui_celular         16650 non-null  int64  
 11  possui_fone_comercial  16650 non-null  int64  
 12  possui_fone            16650 non-null  int64  
 13  possui_email           16650 non-null  int64  
 14  qt_pessoas_residencia  16650 non-null  float64
 15  ma

In [48]:
#tipo dos dados
df.dtypes

sexo                      object
posse_de_veiculo          object
posse_de_imovel           object
qtd_filhos                 int64
tipo_renda                object
educacao                  object
estado_civil              object
tipo_residencia           object
idade                    float64
tempo_emprego            float64
possui_celular             int64
possui_fone_comercial      int64
possui_fone                int64
possui_email               int64
qt_pessoas_residencia    float64
mau                         bool
dtype: object

In [49]:
#quantidade de missings
df.isna().sum()

sexo                     0
posse_de_veiculo         0
posse_de_imovel          0
qtd_filhos               0
tipo_renda               0
educacao                 0
estado_civil             0
tipo_residencia          0
idade                    0
tempo_emprego            0
possui_celular           0
possui_fone_comercial    0
possui_fone              0
possui_email             0
qt_pessoas_residencia    0
mau                      0
dtype: int64

In [50]:
#distribuição da variável resposta 
df['mau'].value_counts()

False    16260
True       390
Name: mau, dtype: int64

#### 2) Vamos montar um metadados

1. Crie um dataframe com os nomes de cada variável e o tipo de dados de cada variável.
2. Adicione uma coluna nesse *dataframe* chamada "qtd_categorias" e coloque nela o número de categorias correspondente de cada variável. 
    Dica: 
        1. inicie uma lista vazia
        2. faça um for ao longo dos nomes das variáveis, 
        3. conte o número de categorias dessa variável 
        4. acumule essa informação de 3. na lista que você criou em 1. 
        5. No final, essa lista pode ser atribuída à nossa variável.
3. Crie variáveis dummy para as variáveis necessárias (i.e. aquelas que são qualitativas e não estão armazenadas como {0, 1} ou {True, False}.

In [51]:
lista_tipo_cat = []
lista_qtd_cat = []

In [52]:
for c in df:
    if df[c].dtype == 'object':
        a = df[c].unique()
        b = df[c].nunique()
        lista_tipo_cat.append(a)
        lista_qtd_cat.append(b)
print(lista_tipo_cat)
print(lista_qtd_cat)
    

[array(['M', 'F'], dtype=object), array(['Y', 'N'], dtype=object), array(['Y', 'N'], dtype=object), array(['Working', 'Commercial associate', 'State servant', 'Pensioner',
       'Student'], dtype=object), array(['Secondary / secondary special', 'Higher education',
       'Incomplete higher', 'Lower secondary', 'Academic degree'],
      dtype=object), array(['Married', 'Single / not married', 'Civil marriage', 'Widow',
       'Separated'], dtype=object), array(['House / apartment', 'Rented apartment', 'Municipal apartment',
       'With parents', 'Co-op apartment', 'Office apartment'],
      dtype=object)]
[2, 2, 2, 5, 5, 5, 6]


In [53]:
df = pd.get_dummies(df, columns= ['tipo_renda', 'educacao', 'estado_civil', 'tipo_residencia'], drop_first = True)
df.head()

sexo posse_de_veiculo posse_de_imovel  qtd_filhos      idade  tempo_emprego  \
0    M                Y               Y           0  58.832877       3.106849   
1    F                N               Y           0  52.356164       8.358904   
2    F                N               Y           0  52.356164       8.358904   
3    M                Y               Y           0  46.224658       2.106849   
4    F                Y               N           0  29.230137       3.021918   

   possui_celular  possui_fone_comercial  possui_fone  possui_email  ...  \
0               1                      0            0             0  ...   
1               1                      0            1             1  ...   
2               1                      0            1             1  ...   
3               1                      1            1             1  ...   
4               1                      0            0             0  ...   

   educacao_Secondary / secondary special  estado_civil_Married  \
0                                       1                     1   
1                                       1                     0   
2                                       1                     0   
3                                       0                     1   
4                                       0                     1   

   estado_civil_Separated  estado_civil_Single / not married  \
0                       0                                  0   
1                       0                                  1   
2                       0                                  1   
3                       0                                  0   
4                       0                                  0   

   estado_civil_Widow  tipo_residencia_House / apartment  \
0                   0                                  1   
1                   0                                  1   
2                   0                                  1   
3                   0                                  1   
4                   0                                  1   

   tipo_residencia_Municipal apartment  tipo_residencia_Office apartment  \
0                                    0                                 0   
1                                    0                                 0   
2                                    0                                 0   
3                                    0                                 0   
4                                    0                                 0   

   tipo_residencia_Rented apartment  tipo_residencia_With parents  
0                                 0                             0  
1                                 0                             0  
2                                 0                             0  
3                                 0                             0  
4                                 0                             0  

[5 rows x 29 columns]

#### 3) Crie variáveis dummy para as variáveis necessárias (i.e. aquelas que são qualitativas e não estão armazenadas como {0, 1} ou {True, False}. Crie um *dataframe* apenas com as variáveis apropriadas para entrada no scikitlearn - elimine as variáveis tipo *str*, mantendo apenas suas versões *dummy*.

In [54]:
df.sexo = df.sexo.map({'F':1, 'M':0})
df.posse_de_veiculo = df.posse_de_veiculo.map({'Y':1, 'N':0})
df.posse_de_imovel = df.posse_de_imovel.map({'Y':1, 'N':0})
df.head()

sexo  posse_de_veiculo  posse_de_imovel  qtd_filhos      idade  \
0     0                 1                1           0  58.832877   
1     1                 0                1           0  52.356164   
2     1                 0                1           0  52.356164   
3     0                 1                1           0  46.224658   
4     1                 1                0           0  29.230137   

   tempo_emprego  possui_celular  possui_fone_comercial  possui_fone  \
0       3.106849               1                      0            0   
1       8.358904               1                      0            1   
2       8.358904               1                      0            1   
3       2.106849               1                      1            1   
4       3.021918               1                      0            0   

   possui_email  ...  educacao_Secondary / secondary special  \
0             0  ...                                       1   
1             1  ...                                       1   
2             1  ...                                       1   
3             1  ...                                       0   
4             0  ...                                       0   

   estado_civil_Married  estado_civil_Separated  \
0                     1                       0   
1                     0                       0   
2                     0                       0   
3                     1                       0   
4                     1                       0   

   estado_civil_Single / not married  estado_civil_Widow  \
0                                  0                   0   
1                                  1                   0   
2                                  1                   0   
3                                  0                   0   
4                                  0                   0   

   tipo_residencia_House / apartment  tipo_residencia_Municipal apartment  \
0                                  1                                    0   
1                                  1                                    0   
2                                  1                                    0   
3                                  1                                    0   
4                                  1                                    0   

   tipo_residencia_Office apartment  tipo_residencia_Rented apartment  \
0                                 0                                 0   
1                                 0                                 0   
2                                 0                                 0   
3                                 0                                 0   
4                                 0                                 0   

   tipo_residencia_With parents  
0                             0  
1                             0  
2                             0  
3                             0  
4                             0  

[5 rows x 29 columns]

In [55]:
df.sexo.astype(int)
df.posse_de_veiculo.astype(int)
df.posse_de_imovel.astype(int)
df.qt_pessoas_residencia.astype(int)
df.mau.astype(int)
df.dtypes

sexo                                        int64
posse_de_veiculo                            int64
posse_de_imovel                             int64
qtd_filhos                                  int64
idade                                     float64
tempo_emprego                             float64
possui_celular                              int64
possui_fone_comercial                       int64
possui_fone                                 int64
possui_email                                int64
qt_pessoas_residencia                     float64
mau                                          bool
tipo_renda_Pensioner                        uint8
tipo_renda_State servant                    uint8
tipo_renda_Student                          uint8
tipo_renda_Working                          uint8
educacao_Higher education                   uint8
educacao_Incomplete higher                  uint8
educacao_Lower secondary                    uint8
educacao_Secondary / secondary special      uint8


#### 4) Qual variável é mais poderosa?

Considere as variáveis ```possui_email``` e ```posse_de_veiculo```. Faça uma tabela cruzada entre elas e responda qual delas te parece mais poderosa para prever a probabilidade de ```mau = 1```?

In [56]:
df[['possui_email', 'posse_de_veiculo', 'mau']].groupby(['mau','posse_de_veiculo']).count()

possui_email
mau   posse_de_veiculo              
False 0                         9931
      1                         6329
True  0                          247
      1                          143

In [57]:
df[['possui_email', 'posse_de_veiculo', 'mau']].groupby(['mau','possui_email']).count()

posse_de_veiculo
mau   possui_email                  
False 0                        14810
      1                         1450
True  0                          360
      1                           30

#### A variável *posse_de_veículo* é mais poderosa de determinar se a pessoa será um *MAU* pagador pois esta variável apresenta em sua descrição que do total de MAUS pagadores (390), 360 nao possuem carro e apenas 30 possuem, ja observando a variável *possui_email* ela se mostra mais equilibrada nao se mostrando uma boa escolha para a predição da variável resposta

#### 5) Salve a base, pois ela será utilizada no final deste módulo.

In [58]:
df.to_csv('base_mod_7.csv')